Code/Framework/AzFramework/Platform/Linux/AzFramework/Input/Devices/Keyboard/InputDeviceKeyboard_xcb.cpp

```diff
@@ -0,0 +1,265 @@
+    {
+    public:
+        AZ_CLASS_ALLOCATOR(InputDeviceKeyboardXcb, AZ::SystemAllocator, 0);
+
+        using InputDeviceKeyboard::Implementation::Implementation;
+        InputDeviceKeyboardXcb(InputDeviceKeyboard& inputDevice)
+            : InputDeviceKeyboard::Implementation(inputDevice)
+        {
+            RawInputNotificationBusXcb::Handler::BusConnect();
+
+            auto* connection = AzFramework::LinuxXcbConnectionManagerInterface::Get()->GetXcbConnection();
+
+            AZStd::unique_ptr<xcb_xkb_use_extension_reply_t, DeleterForFreeFn<::std::free>> xkbUseExtensionReply{
+                xcb_xkb_use_extension_reply(connection, xcb_xkb_use_extension(connection, 1, 0), nullptr)
+            };
+            m_coreDeviceId = xkb_x11_get_core_keyboard_device_id(connection);
+
+            m_xkbContext.reset(xkb_context_new(XKB_CONTEXT_NO_FLAGS));
+            m_xkbKeymap.reset(xkb_x11_keymap_new_from_device(m_xkbContext.get(), connection, m_coreDeviceId, XKB_KEYMAP_COMPILE_NO_FLAGS));
+            m_xkbState.reset(xkb_x11_state_new_from_device(m_xkbKeymap.get(), connection, m_coreDeviceId));
+        }
+
+        bool IsConnected() const override
+        {
+            auto* connection = AzFramework::LinuxXcbConnectionManagerInterface::Get()->GetXcbConnection();
+            return connection && !xcb_connection_has_error(connection);
+        }
+
+        bool HasTextEntryStarted() const override
+        {
+            return false;
+        }
+
+        void TextEntryStart(const InputDeviceKeyboard::VirtualKeyboardOptions& options) override
+        {
+        }
+
+        void TextEntryStop() override
+        {
+        }
+
+        void TickInputDevice() override
+        {
+            ProcessRawEventQueues();
+        }
+
+        void OnRawInputEvent(xcb_generic_event_t* event) override
+        {
+            switch (event->response_type & ~0x80)
+            {
+            case XCB_KEY_PRESS:
+            {
+                auto* keyPress = reinterpret_cast<xcb_key_press_event_t*>(event);
+
+                const InputChannelId* key = InputChannelFromKeyEvent(keyPress->detail);
+                if (key)
+                {
+                    QueueRawKeyEvent(*key, true);
+                }
+                break;
+            }
+            case XCB_KEY_RELEASE:
+            {
+                auto* keyRelease = reinterpret_cast<xcb_key_release_event_t*>(event);
+
+                const InputChannelId* key = InputChannelFromKeyEvent(keyRelease->detail);
+                if (key)
+                {
+                    QueueRawKeyEvent(*key, false);
+                }
+                break;
+            }
+            }
+        }
+
+    private:
+        [[nodiscard]] const InputChannelId* InputChannelFromKeyEvent(xcb_keycode_t code) const
+        {
+            const xcb_keysym_t keysym = xkb_state_key_get_one_sym(m_xkbState.get(), code);
+
+            switch(keysym)
```

Comment:
Probably doesn't make all that much difference in practice, but for Windows and Mac we map the 'raw' keycodes to their corresponding InputChannelId using a fixed size AZStd::array (see Code/Framework/AzFramework/AzFramework/Input/Devices/Keyboard/InputDeviceKeyboardWindowsScanCodes.h and Code/Framework/AzFramework/Platform/Mac/AzFramework/Input/Devices/Keyboard/InputDeviceKeyboard_Mac.mm) which may be slightly more efficient than a big switch statement?

Reply:
From what I see on [godbolt.org](https://godbolt.org/z/3s8WM5c4x), in an optimized build, MSVC and Clang both generate a branch table for this switch statement, and then just jump straight to the right label in constant time.

Reply:
Cool, sounds good. I still slightly prefer using the fixed array simply for consistency with the implementations on other platforms, but no worries at all if you prefer to keep the switch statement.